lib/aot/compiler.cpp

```diff
@@ -2905,6 +3145,145 @@ class FunctionCompiler {
+
+    Builder.SetInsertPoint(OkBB);
+  }
+
+  void compileMemoryFence() {
+    Builder.CreateFence(llvm::AtomicOrdering::SequentiallyConsistent);
+  }
+  void compileAtomicNotify() {
+    Builder.CreateCall(Context.getIntrinsic(
+                           Builder,
+                           AST::Module::Intrinsics::kMemoryAtomicNotify,
+                           llvm::FunctionType::get(Context.VoidTy, {}, false)),
+                       {});
+  }
+  void compileAtomicWait(unsigned MemIdx, unsigned BitWidth) {
+    auto *Offset = Builder.CreateSExtOrTrunc(stackPop(), Context.Int32Ty);
+    auto *ExpectedValue =
+        Builder.CreateSExtOrTrunc(stackPop(), Context.Int64Ty);
+    auto *Timeout = Builder.CreateSExtOrTrunc(stackPop(), Context.Int64Ty);
+
+    stackPush(Builder.CreateCall(
+        Context.getIntrinsic(
+            Builder, AST::Module::Intrinsics::kMemoryAtomicWait,
+            llvm::FunctionType::get(Context.Int32Ty,
+                                    {Context.Int32Ty, Context.Int32Ty,
+                                     Context.Int64Ty, Context.Int64Ty,
+                                     Context.Int32Ty},
+                                    false)),
+        {Builder.getInt32(MemIdx), Offset, ExpectedValue, Timeout,
+         Builder.getInt32(BitWidth)}));
+  }
+  void compileAtomicLoad(unsigned MemoryIndex, llvm::IntegerType *IntType,
+                         llvm::IntegerType *TargetType, bool Signed = false) {
+
+    auto *Offset = Builder.CreateZExt(Stack.back(), Context.Int64Ty);
+    compileAtomicCheckOffsetAlignment(Offset, TargetType);
+    auto *VPtr = Builder.CreateInBoundsGEP(
+        Context.Int8Ty, Context.getMemory(Builder, ExecCtx, MemoryIndex),
+        Offset);
+
+    auto *Ptr = Builder.CreateBitCast(VPtr, TargetType->getPointerTo());
+    auto *Load = Builder.CreateLoad(TargetType, Ptr, OptNone);
+    Load->setAlignment(Align(8));
+    Load->setAtomic(llvm::AtomicOrdering::SequentiallyConsistent);
+
+    if (Signed) {
+      Stack.back() = Builder.CreateSExt(Load, IntType);
+    } else {
+      Stack.back() = Builder.CreateZExt(Load, IntType);
+    }
+  }
+  void compileAtomicStore(unsigned MemoryIndex,
+                          [[maybe_unused]] llvm::IntegerType *IntType,
+                          llvm::IntegerType *TargetType, bool Signed = false) {
+
+    auto *V = stackPop();
+
+    if (Signed) {
+      V = Builder.CreateSExtOrTrunc(V, TargetType);
+    } else {
+      V = Builder.CreateZExtOrTrunc(V, TargetType);
+    }
+    auto *Offset = Builder.CreateZExt(Stack.back(), Context.Int64Ty);
+    compileAtomicCheckOffsetAlignment(Offset, TargetType);
+    auto *VPtr = Builder.CreateInBoundsGEP(
+        Context.Int8Ty, Context.getMemory(Builder, ExecCtx, MemoryIndex),
+        Offset);
+    auto *Ptr = Builder.CreateBitCast(VPtr, TargetType->getPointerTo());
+    auto *Store = Builder.CreateStore(V, Ptr, OptNone);
+    Store->setAlignment(Align(8));
```

Comment:
You need to get alignment parameter from `Instruction` for llvm generating proper atomic instruction.